### PR TITLE
Update draft-irtf-cfrg-randomness-improvements.md

### PR DESCRIPTION
**Pull request by smyshsv**
Addressing issue #31 .

(And one minor misprint)

### CODE REVIEW

File: draft-irtf-cfrg-randomness-improvements.md [hunk 1]
```diff
@@ -214,7 +224,7 @@ To provide security in the cases of usage of CSPRNGs in virtual environments,
 it is RECOMMENDED to incorporate all available information specific to the process that 
 would ensure the uniqueness of each tag1 value among different instances of virtual machines 
 (including ones that were cloned or recovered from snapshots). 
-It is needed to address the problem of CSPRNG state cloning (see {{RY2010}}.
+It is needed to address the problem of CSPRNG state cloning (see {{RY2010}}).
```
Comment (chris-wood) [inline draft-irtf-cfrg-randomness-improvements.md]: 👍

File: draft-irtf-cfrg-randomness-improvements.md [hunk 2]
```diff
@@ -198,7 +204,11 @@ or use an independent (and completely reliable) entropy source, e.g., if Sig is
 in an HSM with its own internal trusted entropy source for signature generation.
 
 In systems where signature computations are expensive, G'(n) may be precomputed and pooled.
-This is possible since the construction depends solely upon the CSPRNG output and private key.
+This is possible since the construction depends solely upon the CSPRNG output and private key. 
+Experimental results indicate that caching the signature output is critical for performance. 
+The signature computation cost is amortized in about 1000 calls, afterwards each call of G’(n) 
+instead of G(n) costs from nanoseconds to microseconds (depending on the required output size) 
+-- i.e., the relative cost is minor with respect to cryptographic operations in protocols such as TLS.
```
Comment (chris-wood) [inline draft-irtf-cfrg-randomness-improvements.md]: s/minor/negligible?

File: draft-irtf-cfrg-randomness-improvements.md [hunk 3]
```diff
@@ -198,7 +204,11 @@ or use an independent (and completely reliable) entropy source, e.g., if Sig is
 in an HSM with its own internal trusted entropy source for signature generation.
 
 In systems where signature computations are expensive, G'(n) may be precomputed and pooled.
-This is possible since the construction depends solely upon the CSPRNG output and private key.
+This is possible since the construction depends solely upon the CSPRNG output and private key. 
+Experimental results indicate that caching the signature output is critical for performance. 
+The signature computation cost is amortized in about 1000 calls, afterwards each call of G’(n)
```
Comment (chris-wood) [inline draft-irtf-cfrg-randomness-improvements.md]: This number probably doesn't mean much without a description of the implementation and experimental platform. Can we move this bit to a separate PR where we can address that in full?

File: draft-irtf-cfrg-randomness-improvements.md (hunk 2 above)
Comment (smyshsv) [reply draft-irtf-cfrg-randomness-improvements.md]: Yes, you're right!

File: draft-irtf-cfrg-randomness-improvements.md (hunk 3 above)
Comment (smyshsv) [reply draft-irtf-cfrg-randomness-improvements.md]: I agree that mentioning "about 1000" without any information about the experiments may look quite lame. My opinion here was that a particular number is not necessary - only the idea of "cache the signature, or it will be really slower" should be reflected.

But you are right that it doesn't look good. I'll remove this one from this PR - let's discuss, how to add details about the experiments separately (taking into account adding an explicit reference to your paper with experiments).

File: draft-irtf-cfrg-randomness-improvements.md [hunk 4]
```diff
@@ -197,8 +203,14 @@ Sig MUST be a deterministic signature function, e.g., deterministic ECDSA {{RFC6
 or use an independent (and completely reliable) entropy source, e.g., if Sig is implemented 
 in an HSM with its own internal trusted entropy source for signature generation.
 
-In systems where signature computations are expensive, G'(n) may be precomputed and pooled.
-This is possible since the construction depends solely upon the CSPRNG output and private key.
+In systems where signature computations are expensive, Sig(sk, tag1) may be cached. 
+Experimental results indicate that caching the signature output is critical for performance --
```
Comment (chris-wood) [inline draft-irtf-cfrg-randomness-improvements.md]: I suggest we remove this sentence and address it in a separate PR. Other than that, this is good!
